pybamm/models/submodels/interface/open_circuit_potential/plett_ocp.py

```diff
@@ -0,0 +1,135 @@
+                h = pybamm.PrimaryBroadcast(h, [f'{domain} particle size'])
+            else:
+                sto_surf = variables[
+                    f"{Domain} {phase_name}particle surface stoichiometry"
+                ]
+                # If variable was broadcast, take only the orphan
+                if isinstance(sto_surf, pybamm.Broadcast) and isinstance(
+                    T, pybamm.Broadcast
+                ):
+                    sto_surf = sto_surf.orphans[0]
+                    T = T.orphans[0]
+
+            ocp_surf_eq = self.phase_param.U(sto_surf, T)
+            dUdT = self.phase_param.dUdT(sto_surf)
+
+            # Bulk OCP is from the average SOC and temperature
+            sto_bulk = variables[f"{Domain} electrode {phase_name}stoichiometry"]
+            T_bulk = pybamm.xyz_average(pybamm.size_average(T))
+            ocp_bulk_eq = self.phase_param.U(sto_bulk, T_bulk)
+
+            c_s_rav = variables[
+            f"R-averaged {domain} {phase_name}particle concentration distribution"
+            ]
+            # eps_s = variables[
+            #     f"{Domain} electrode {phase_name}active material volume fraction"
+            # ]
+            # eps_s_av = pybamm.x_average(eps_s)
+            # c_s_vol_av = pybamm.x_average(eps_s * c_s_rav) / eps_s_av
+            # c_s_vol = eps_s * c_s_rav
+            c_scale = self.phase_param.c_max
+            # L = self.domain_param.L
+            # A = self.param.A_cc
+            sto_rav = c_s_rav / c_scale
+            sto_rav = variables[f'R-averaged {domain} particle stoichiometry distribution']
+            # sto_xav = variables[f'X-averaged {domain} particle stoichiometry distribution']
+            # eps_s = pybamm.PrimaryBroadcast(eps_s,[f'{domain} particle size'])
+
+            #! don't know how to broadcast this
+            variables[f"Total lithium in {phase} phase in {domain} electrode [mol]"] =  sto_bulk* c_scale #c_s_vol * L * A
+
+
+            H = self.phase_param.H(sto_rav)
+            variables[f'{Domain} electrode {phase_name}equilibrium OCP [V]'] = ocp_surf_eq
+            variables[f'{Domain} electrode {phase_name}bulk equilibrium OCP [V]'] = ocp_bulk_eq
+            variables[f'{Domain} electrode {phase_name}OCP hysteresis [V]'] = H
+            # dQdU = variables[f'{Domain} electrode {phase_name}differential capacity [A.s.V-1]']
+
+            dU = pybamm.source(1,ocp_bulk_eq.diff(sto_bulk))
+            # dU = pybamm.PrimaryBroadcast(dU,[f'{domain} electrode'])
+            # dU = pybamm.PrimaryBroadcast(dU,[f'{domain} particle size'])
+            # dU = pybamm.FullBroadcast(dU,broadcast_domains={'primary':f'{domain} particle size', 'secondary':f'{domain} electrode'})
+            Q = variables[f"Total lithium in {phase} phase in {domain} electrode [mol]"] * pybamm.constants.F / 3600
+            # Q = pybamm.PrimaryBroadcast(Q,f'{domain} electrode')
+            dQ = Q.diff(sto_bulk)
```

Comment:
@tinosulzer I am pretty sure that the correct domains for the dU  and  dQ variables should be  the current collector and the electrode. In this case, I am trying to differentiate Q by sto_bulk, although I often end up with a scalar 0 after. Are there specific rules or best practices for differentiating?

Reply:
I finished implementing the dU calculation